py3status/modules/mysql_slave.py

```diff
@@ -0,0 +1,86 @@
+# -*- coding: utf-8 -*-
+"""
+Display MySQL/MariaDB slave state.
+
+This module displays the number of seconds the slave is behind the master.
+
+Configuration parameters:
+    cache_timeout: refresh cache_timeout for this module (default 5)
+    format_master: format used for master_servers
+                    (default '[\\?color=seconds {host} is master]')
+    format_slave: format used for slave servers
+                    (default '[\\?color=seconds {host} is {seconds}s behind]')
+    host: the host to connect to (default None)
+    passwd: the password to use for the connection (default None)
+    port: the port to connect to (default 3306)
+    thresholds: a threshold list of tuples, -1 used for a master hosts
+      (default [(-1, 'blue'), (0, 'deepskyblue'), (100, 'good'), (300, 'degraded'), (600, 'bad')])
+    user: the user used for the connection (default None)
+
+Requires:
+    MySQLdb: python package for python
+
+@author cereal2nd
+@license BSD
+
+SAMPLE OUTPUT
+{'full_text': 'test is 0s behind'}
+{'full_text': 'test is 509s behind'}
+{'full_text': 'test is master'}
+"""
+
+from __future__ import absolute_import
+from MySQLdb import connect
+
+
+class Py3status:
+    """
+    """
+    # available configuration parameters
+    cache_timeout = 5
+    format_master = '[\?color=seconds {host} is master]'
+    format_slave = '[\?color=seconds {host} is {seconds}s behind]'
+    host = None
+    passwd = None
+    port = 3306
+    thresholds = [
+            (-1, 'blue'),
+            (0, 'deepskyblue'),
+            (100, 'good'),
+            (300, 'degraded'),
+            (600, 'bad')
+        ]
+    user = None
+
+    def mysql_slave(self):
+        try:
+            data = self._get_slave_status()
+            seconds = data['Seconds_Behind_Master']
+            form = self.format_slave
+        except:
+            form = self.format_master
+            seconds = -1
+        self.py3.threshold_get_color(seconds, 'seconds')
+
+        return {
+            'cached_until': self.py3.time_in(self.cache_timeout),
+            'full_text': self.py3.safe_format(form, {'seconds': str(seconds), 'host': self.host})
+        }
+
+    def _get_slave_status(self):
```

Comment:
(My QA) Move this function up and rename this to `_get_(module_name)_data`. Consistency with other modules. Can this be used on master too? If yes, then maybe make it more generic by renaming it to `mysql.py` and make a config for `cur.execute()` command? Would that be more useful for `mysql`?

eg If I only want `Seconds_Behind_Master`, then I can leave the command as-is or customize the command to only query that name instead of all fields (less work). Does that make sense or nah?

Reply:
i does make sense, but what if the query in the cur.execute returns more then one row?
How do we handle this?

If we do this this becomes more a mysql_fetch module

Reply:
I'm not familiar with this. Looking at `thunderbird_todos.py` may help you with some ideas of how it ought to works. I assume we would loop and do stuffs with them.

Reply:
The solution could work if we force the sql query to only return one row.

if we then get back something like
{'foo': '1', 'bar':2, 'x':'y', 'z':'a'}
via format we can indeed display the items we want, but how do we handle the thresholds, we need these per item in this case.

Also if we do something like this the module will be a lot less userfriendely it becomes a more advanced module with a lot of config ...

Reply:
For now... add this:
```
            for k, v in data.items():
                self.py3.threshold_get_color(v, k)
```


Starting with this module doing no more than printing `seconds` behind master, we can make this module more practical by exposing all placeholders, supporting thresholds for all, and whatnot... Then users can query what they want with `mysql`.

For complicated scenarios, we could add examples too. So far, I hadn't seen one.

Reply:
and how can we provide thresholds per placeholder?

Reply:
Ooo. This one.
```
format = '[\?color=seconds_behind_master {seconds_behind_master} seconds] '
format += '[\?color=name1 {name1} ] '
format += '[\?color=name2 {name2} ] '
format += '[\?color=name3 {name3} ] '
```

It does sucks that MySQL gives us long placeholders though.

Reply:
The placeholder names are basically the column names in this case.

I was more talking about how do we define thresholds (in the config) per placeholder ...

      thresholds = [
              (-1, 'blue'),                                                               
              (0, 'deepskyblue'),                                                         
              (100, 'good'),
              (300, 'degraded'),                                                          
              (600, 'bad')
          ]  

currently thresholds is a list, can this also be a dict?

Reply:
thresholds = {
        'seconds_behind_master': [
            (-1, 'blue'),
            (0, 'deepskyblue'),
            (100, 'good'),
            (300, 'degraded'),
            (600, 'bad')
        ],
        'name1': [(0, 'good'), (1, 'degraded'), (2, 'bad')],
        'name2': [(0, 'good'), (1, 'degraded'), (2, 'bad')],
        'name1': [(0, 'good'), (1, 'degraded'), (2, 'bad')],
    }